### PR TITLE
feat: Support Ref Cleanup

### PR DESCRIPTION
**Pull request by chirokas**
Closes https://github.com/adobe/react-spectrum/issues/8024

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/WAI/ARIA/apg/)

### CODE REVIEW

**Comment by snowystinger** [comment]
Thanks for the interest, it appears this is a React 19 only feature https://react.dev/blog/2024/12/05/react-19#cleanup-functions-for-refs

I noticed that all the tests, React 16, 17 and 18, pass. How much of the work here makes it work with previous versions? I still get a console warning if I try to use it, so I don't think it's a use case we should support. We can turn off the test for the other versions of React.

**Comment by chirokas** [comment]
> I still get a console warning if I try to use it

```tsx
// React 16, 17 and 18

<input ref={mergeRefs(onRefChangeWithCleanup)} />
// Warning: Unexpected return value from a callback ref
// https://github.com/facebook/react/pull/22313
```

We don't use ref cleanups internally, so this still happens in React < 19 when a user's ref callback returns a value.

However, it works well

```tsx
<input ref={useObjectRef(mergeRefs(onRefChangeWithCleanup))} />
```

**Review by snowystinger — APPROVED** [review]
Initially looks good, I've added some preliminary mergeRef tests in case we think of others edge cases we should try

File: packages/@react-aria/utils/src/useObjectRef.ts [hunk 1]
```diff
@@ -10,30 +10,60 @@
...
+      },
+      set current(value) {
+        objRef.current = value;
+        if (cleanupRef.current) {
+          cleanupRef.current();
+          cleanupRef.current = undefined;
+        }
+
+        if (value != null) {
+          cleanupRef.current = refEffect(value);
```
Comment (devongovett) [inline packages/@react-aria/utils/src/useObjectRef.ts:63]: We might be missing a call to cleanup on unmount. I guess we'd need to add a useEffect cleanup to trigger that. Would the order difference between real ref setup/cleanup and effect setup/cleanup break in strict mode?

File: packages/@react-aria/utils/src/useObjectRef.ts [hunk 2]
```diff
@@ -10,30 +10,60 @@
...
+  const refEffect = useCallback(
+    (instance: T | null) => {
+      if (typeof ref === 'function') {
+        const refCallback = ref;
+        const refCleanup = refCallback(instance);
+        return () => {
+          if (typeof refCleanup === 'function') {
+            refCleanup();
+          } else {
+            refCallback(null);
```
Comment (devongovett) [inline packages/@react-aria/utils/src/useObjectRef.ts:37]: Probably not important, but there's a slight behavior change here:

```js
let ref = useObjectRef(value => {});
ref.current = {};
ref.current = undefined
```

Previously, the function would be called twice with {} and undefined. Now it will be called with {} and null.

**Review by devongovett — APPROVED** [review]

File: packages/@react-aria/utils/src/useObjectRef.ts [hunk 3]
```diff
@@ -10,30 +10,60 @@
...
+        return objRef.current;
+      },
+      set current(value) {
+        objRef.current = value;
+        if (cleanupRef.current) {
+          cleanupRef.current();
+          cleanupRef.current = undefined;
+        }
+
+        if (value != null) {
```
Comment (devongovett) [inline packages/@react-aria/utils/src/useObjectRef.ts:62]: Also this means there's no way to distinguish between `null` to trigger a cleanup vs `null` being a valid value of the ref. But not sure there's a way around that.

**Comment by devongovett** [comment]
Aiming for a release next week.
